Refactor text trimming and implement prefix trimming.

MarchingCube opened the pull request:
## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
Current text trimming setup was fairly limiting and didn't allow users to customize behavior (for example using custom ellipsis symbols) or parametrize trimming.

I'm also adding new kind of trimming that could be found in programs like Adobe After Effects (trimming that keeps a fixed prefix and has ellipsis in the middle). Such trimming is useful in cases where prefix and suffix is more important than content in the middle.

This trimming mode is configurable so new changed API is immediately useful here.

https://user-images.githubusercontent.com/2588062/148532342-a1b57183-8c0c-4aa0-b9b2-2f25dfa60890.mp4

I've implemented XAML compiler intrinsic handling for both text trimming and decorations. Previously if one used `TextDecorations="Underline"` it would create a completely new collection of decorations.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
One cannot easily configure trimming behavior and only trailing word/character trimming is available.

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
Users can create custom trimming configuration and use it with `TextBlock`. One can enable new `PrefixEllipsis` to see new trimming in action.

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->
Using an enum is not enough to configure trimming fully so instead it is a class. I've added static properties that have matching names with old enum members, this is comparable to `TextDecorations`.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
<!--- List any breaking changes here. When the PR is merged please add an entry to https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes -->

`TextTrimming` is no longer an enum. This breaks binary compatibility. XAML files should compile as is, most of usages in code will work too. Only cases when user expected `TextTrimming` to be an enum won't work.

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

robloo commented:
I was worried when I saw the TextTrimming enum being dropped -- there is a lot of existing code that uses this both in XAML and C#. This is a smartly designed change though and it looks easy to integrate into existing apps. The new PrefixEllipsis is very welcome and has been missing, well, forever in XAML. It's a great idea!

Using "This is the longest name" as an example it seems the following is possible:

Word: "This is the longest..."
Character: "This is the longest na..."
PrefixEllipsis[4]: "This...longest name"

What about the case where only the beginning should be trimmed with an ellipse? This is useful for file paths and other types of text where most significant info is at the end. In WPF there was no way to do this: "... the longest name". I'm assuming it can be done now with PrefixEllipsis and the length set to zero? Just want to be sure that case is fully supported.

Then I wonder if actually PrefixEllipsis with length set to zero should be the default instead of an arbitrary length of 8 characters. This would make it symmetrical by default with character trimming at the end.

Finally, can the PrefixEllipsis length be set in XAML? If so, what is the syntax? The docs really should be updated for that.

timunie commented:
Hi,

Imo this is an interesting idea. I have two questions/suggestions if you don't mind:

1. What if the important content is at the beginning, but one wants also to keep the last lets say 4 chars? Would that be possible?  
2. A readonly bool IsTrimmed may be a good addition. So the developers could provide a tooltip if trimmed, and only if trimmed. 

Happy coding
Tim

Gillibald commented:
> Hi,
> 
> Imo this is an interesting idea. I have two questions/suggestions if you don't mind:
> 
> 1. What if the important content is at the beginning, but one wants also to keep the last lets say 4 chars? Would that be possible?
> 2. A readonly bool IsTrimmed may be a good addition. So the developers could provide a tooltip if trimmed, and only if trimmed.
> 
> Happy coding Tim

Trimming happens on TextLine level and for collapsed lines this is set https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Media/TextFormatting/TextLine.cs#L56

Gillibald commented:
Maybe it might make sense to introduce a position property on TextTrimming so we are able to chose on which side the text is trimmed. PrefixLength will become OmittedLength or something.

kekekeks commented on src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguageParseIntrinsics.cs:
```diff
@@ -221,6 +221,32 @@ public static bool TryConvert(AstTransformationContext context, IXamlAstValueNod
                 }
             }
 
+            if (type.Equals(types.TextTrimming))
```
IMO we should add `[XamlEnum]` and `[XamlEnumMember]` attributes and look for those instead to make it more generic. Also easier for IDE's to add support for

timunie commented:
> > Hi,
> > 
> > Imo this is an interesting idea. I have two questions/suggestions if you don't mind:
> > 
> > 1. What if the important content is at the beginning, but one wants also to keep the last lets say 4 chars? Would that be possible?
> > 2. A readonly bool IsTrimmed may be a good addition. So the developers could provide a tooltip if trimmed, and only if trimmed.
> > 
> > Happy coding Tim
> 
> Trimming happens on TextLine level and for collapsed lines this is set https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Media/TextFormatting/TextLine.cs#L56


Thank you for your detailed reply

So I could check if this property is true for any line, if I want to know if any trimming happened? 


```cs
public​ ​abstract​ ​bool​ ​HasOverflowed​ { ​get​; }
```

Gillibald commented:
> > > Hi,
> > > Imo this is an interesting idea. I have two questions/suggestions if you don't mind:
> > > 
> > > 1. What if the important content is at the beginning, but one wants also to keep the last lets say 4 chars? Would that be possible?
> > > 2. A readonly bool IsTrimmed may be a good addition. So the developers could provide a tooltip if trimmed, and only if trimmed.
> > > 
> > > Happy coding Tim
> > 
> > 
> > Trimming happens on TextLine level and for collapsed lines this is set https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Media/TextFormatting/TextLine.cs#L56
> 
> Thank you for your detailed reply
> 
> So I could check if this property is true for any line, if I want to know if any trimming happened?
> 
> ```cs
> public​ ​abstract​ ​bool​ ​HasOverflowed​ { ​get​; }
> ```

You can look for lines that have `HasCollapsed` set to true

timunie commented:
Thank you @Gillibald. Now I got it. 👍

robloo commented:
I realize we probably need a separate spec discussion for this. Fundamentally, there are a few different generalized concepts:

 * BreakMode : By 'Word' or 'Character'
 * EllipsisLocation : Start (new), End [Note that Middle is not required as the PrefixLength/SuffixLength imply this instead]
 * PrefixLength/SuffixLength (new) [When used, these position the ellipsis in the middle of the text]

This all could and probably should be calculated using a single generalized method which would support:

| Break | Location | Example | |
|--------|----------|-----------|-|
| Word | End | "first second ..." |
| Word | Start | "... second third" |
| Character | End | "first second th..." | SuffixLength=0 |
| Character | End | "first seco...hird" | SuffixLength=4 |
| Character | Start | "...st second third" | PrefixLength=0 |
| Character | Start | "firs...cond third" | PrefixLength=4 |

Character break mode obviously supports a configurable PrefixLength but should be extended to support a SuffixLength as well as commented above. This would allow ellipsis in the middle that are fully configurable.

Open Questions:
 1. How to make PrefixLength and SuffixLength configurable in XAML
 2. Rules for Left-to-right and symbolic languages (without a distinction between words/characters). Although I can't think of any special cases needed.
 3. Defaults for PrefixLength and SuffixLength, strong suggestion is zero for both

Gillibald commented:
Under CSS there is a spec for this without a prefix length: https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow

It is possible to trim text at both sides and also define a custom symbol individually. So we could adapt this syntax but need to add the Prefix/SuffixLength

robloo commented:
CSS has some good ideas like making the overflow indicator a configurable string. That could also be implemented relatively easily in the underlying functionality here.

I will also amend the table above. PrefixLength/SuffixLength can really be the same property. It can also apply to word break mode.

jmacato commented:
Closing due to inactivity and merge conflicts. @MarchingCube reopen this if it's gonna be worked on again :)

Gillibald commented:
Will resolve conflicts in a few hours

robloo commented:
I don't think this is quite general-purpose enough. Merging as-is may very well be a restriction in the future.

This is brand-new functionality. API and features should be well discussed IMO.

Gillibald commented:
Had a quick look at this and it seems master did diverge too much recently. We might need to recreate everything on top of master.

Gillibald commented:
@robloo We can add more advanced trimming modes after this PR got merged. This is mainly aiming on being able to customize the text trimming by replacing the TextTrimming enum with a class that is responsible for TextLine collapsing.

robloo commented:
> @robloo We can add more advanced trimming modes after this PR got merged. This is mainly aiming on being able to customize the text trimming by replacing the TextTrimming enum with a class that is responsible for TextLine collapsing.

Alright, as long as the API isn't locked after this. Some of my comments about naming of the new class members should be considered at least though.

MarchingCube committed:
Refactor text trimming and implement prefix trimming.

MarchingCube commented:
@Gillibald I have found some time to resolve conflicts. Since text formatting changes are mostly wrapped up we can revisit this PR again.

I've added a TODO to both ellipsis functions since I'm not dealing with flow direction yet. Also I've added `TryMeasureCharactersBackwards` as a temporary placeholder until we figure API.

If prefix trimming is too controversial to include now we can also look into text formatting API to let users implement it externally. Since I would be fine with moving this implementation to my project until we figure out API surface and how to configure it. In my opinion it should only work with character level trimming since prefix length is given in characters.

Ellipsis symbol is already configurable in my PR (each trimming ctor takes it). Only problem would be XAML syntax although we might be missing a few features in there to make this work (constructor params primarily). Although I would expect users to either create a new static trimming property and reference it via `x:Static`, might be something for another contributor to pick up later.

robloo commented:
>  In my opinion it should only work with character level trimming since prefix length is given in characters.

 1. This isn't matching/symmetrical with the existing `TextTrimming` functionality that supports both word and character (`CharacterEllipsis`, `WordEllipsis`). 
 2. It would be non-obvious that `PrefixEllipse` applies only for character as it isn't part of the name like the other enum values.
 3. There is an opportunity loss if this PR only supports characters. It also restricts future expansion not to account for both `PrefixCharacterEllipsis` and `PrefixWordEllipsis` at least in name.
 4. The prefix `Length` property can be unitless. It doesn't need to represent only characters or words, it can be a unitless integer for both.
 5. A good algorithm would support everything I outlined anyway as it is symmetrical functionality. You are going to have to implement all of the pieces anyway.

My recommendation is to at least:

 1. Change the name from  `PrefixEllipse` -> `PrefixCharacterEllipsis` to allow future expansion without breaking changes
 2. Use a unitless `Length` property that could in the future apply for both character and word trimming

Gillibald commented on src/Avalonia.Visuals/Media/TextCollapsingCreateInfo.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Avalonia.Media
+{
+    public readonly struct TextCollapsingCreateInfo
```
Name suggestion `TextCollapsingSymbol`

Gillibald commented on src/Avalonia.Visuals/Media/TextFormatting/TextEllipsisHelper.cs:
```diff
@@ -0,0 +1,99 @@
+            var textRange = textLine.TextRange;
+
+            // TODO: From where this is supposed to come now?
+            var flowDirection = FlowDirection.LeftToRight;
```
At this point text runs are not reordered so we always have glyphs in one direction

MarchingCube replied:
Added one to `TextCollapsingProperties.Collapse`. Let me know if you would rather have a context struct here or if we should pass paragraph properties instead.

MarchingCube committed:
Allow for passing flow direction when collapsing lines.

MarchingCube committed:
Documentation fixes.

robloo commented on src/Avalonia.Visuals/Media/TextTrimming.cs:
```diff
@@ -1,23 +1,71 @@
+        /// <summary>
+        /// Text is trimmed after a given prefix length. An ellipsis (...) is drawn in between prefix and suffix and represents remaining text.
+        /// </summary>
+        public static TextTrimming PrefixCharacterEllipsis { get; } = new TextLeadingPrefixTrimming(s_defaultEllipsisChar, 8);
```
Why default to 8? Zero seems to make more sense and behaves the inverse of CharacterEllipsis.

MarchingCube replied:
Mostly based on AfterEffects default value (which seems to be 7 or 8) since my implementation idea was from there.
0 works as well although from my perspective it should be called `LeadingCharacterEllipsis` then since with 0 there is no prefix.

MarchingCube replied:
We might change things around when working on css like syntax support for this. Since it would fix problems with trimming permutations and perhaps require a few changes to exposed members. Current API will get refined a bit.

robloo replied:
Ok, good to know there is past precedent then. It's also a good idea to add another "value" for LeadingCharacterEllipsis. That would allow both cases in XAML. Since length cannot be configured in XAML that might actually be better overall.

MarchingCube committed:
Seal APIs.

MarchingCube committed:
Leftover unsealed API.

MarchingCube committed:
API compat.

Gillibald commented on src/Avalonia.Visuals/Media/TextFormatting/TextCollapsingProperties.cs:
```diff
@@ -1,23 +1,27 @@
-        public abstract TextCollapsingStyle Style { get; }
+        /// <param name="textLine">Text line to collapse.</param>
+        /// <param name="flowDirection">Text flow direction.</param>
+        public abstract IReadOnlyList<TextRun>? Collapse(TextLine textLine, FlowDirection flowDirection);
```
Lets just default to FlowDirection.LeftToRight in all cases. Ideally we need to inherit the current run's bidi level but that needs more adjustments I will do later.

MarchingCube committed:
API changes and test fixes.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019298-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

Gillibald committed:
Merge branch 'master' into refactor-text-trimming

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019301-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]
